compiler/rustc_hir_typeck/src/_match.rs

```diff
@@ -283,23 +288,88 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         coercion.coerce_forced_unit(
             self,
             &cause,
             |err| {
-                if let Some((span, msg)) = &ret_reason {
-                    err.span_label(*span, msg.clone());
-                } else if let ExprKind::Block(block, _) = &then_expr.kind
-                    && let Some(expr) = &block.expr
+                if let Some((if_span, msg)) = &ret_reason {
+                    err.span_label(*if_span, msg.clone());
+                } else if let ExprKind::Block(block, _) = then_expr.kind
+                    && let Some(expr) = block.expr
                 {
                     err.span_label(expr.span, "found here");
                 }
                 err.note("`if` expressions without `else` evaluate to `()`");
                 err.help("consider adding an `else` block that evaluates to the expected type");
                 error = true;
+                if let ExprKind::Let(hir::Let { span, pat, init, .. }) = cond_expr.kind
+                    && let ExprKind::Block(block, _) = then_expr.kind
+                    // Refutability checks occur on the MIR, so we approximate it here by checking
+                    // if we have an enum with a single variant or a struct in the pattern.
+                    && let PatKind::TupleStruct(qpath, ..) | PatKind::Struct(qpath, ..) = pat.kind
+                    && let hir::QPath::Resolved(_, path) = qpath
+                {
+                    match path.res {
+                        Res::Def(DefKind::Ctor(CtorOf::Struct, _), _) => {
+                            // Structs are always irrefutable. Their fields might not be, but we
+                            // don't check for that here, it's only an approximation.
+                        }
+                        Res::Def(DefKind::Ctor(CtorOf::Variant, _), def_id)
+                            if self
+                                .tcx
+                                .adt_def(self.tcx.parent(self.tcx.parent(def_id)))
+                                .variants()
+                                .len()
+                                == 1 =>
+                        {
+                            // There's only a single variant in the `enum`, so we can suggest the
+                            // irrefutable `let` instead of `if let`.
+                        }
+                        _ => return,
+                    }
+
+                    let mut sugg = vec![
+                        // Remove the `if`
+                        (if_span.until(*span), String::new()),
+                    ];
+                    match (block.stmts, block.expr) {
+                        ([first, ..], Some(expr)) => {
+                            let padding = self
+                                .tcx
+                                .sess
+                                .source_map()
+                                .indentation_before(first.span)
+                                .unwrap_or_else(|| String::new());
```

Comment:
@wesleywiser yeah. There are a couple of methods to check for "type implements trait?", but they are not always available in all stages. We should probably audit all cases and unify functionality behind a handful of easier to use methods. It is a bit complicated by the needs of diagnostics, where sometimes we want to be maximally inclusive (accepting things that *could* implement, but we don't know for sure) or maximally exclusive (accepting only things that we know for sure implement, to avoid invalid suggestions).